### DISCUSSION TITLE
Support ChatGPT/Open Assistant Plugins via .well-known/ai-plugin.json

### DISCUSSION THREAD

**Discussion by stefangrotz** [ckan/ckan#7501] [Ideas]
ChatGPT is right now starting to roll out a new Plugin system that allows it to access information from external APIs.

Here are the details: 
* https://platform.openai.com/docs/plugins/introduction
* https://platform.openai.com/docs/plugins/getting-started

You basically have to create the file
`yourdomain.com/.well-known/ai-plugin.json`
Where you describe your API in human language. 

It would be cool if admins would be able to easily create a file like this for their portal, because this would enable you to search through a portal using human language instead of SPARQL.

EDIT: Open Assistant is now also supporting this plugin structure.

**Comment by amercader** [comment]
This definitely sounds like an interesting use case for an extension. Although there are concerns about OpenAI business model and practices, ChatGPT plugins seem like the easiest approach to integrate LLMs into CKAN search. I guess that initially that would focus on dataset search, describing the `package_search` endpoint but perhaps there is scope to extending it to queries for the data itself hooking into `datastore_search` (or `datastore_search_sql` if that is enabled). This is a [good article](https://simonwillison.net/2023/Mar/24/datasette-chatgpt-plugin/) for a similar scenario.

**Reply by stefangrotz** [reply]
I share your concerns about OpenAI, but I believe that the /.well-known/ai-plugin.json might become a standard to expose an API to all kinds of AI assistants in the future. I hope free projects like Open Assistant or DuckAssist from DuckDuckGo will work similarly. 

This is now a very early stage, but I belive once the plugin feature of ChatGPT becomes public for more users, the advantages will become obvious. You could directly access data from an Open Data Portal using human language.

At the moment we can focus on learning how this standard works exactly and evaluating it.

**Comment by thegostev** [comment]
Hi, @stefangrotz thanks for the suggestion. Adding ChatGPT to CKAN is a leap forward from UX perspective. ChatGPT model is not a new technology, it was there for years, the latest addition that made it so in demand was the addition of the chatbot. Enquiring data using a human language is a whole new level of accessibility.

Do you have plans to integrate ChatGPT with your portal or have people who'd be interested to test how it works?

If we're working with publicly available data I don't sense issues with privacy, am I missing something?

**Comment by stefangrotz** [comment]
Open Assistant (https://open-assistant.io/) now also supports plugins. I started to experiment with CKAN plugins for a local open data portal using these two files:

* https://stefang.uber.space/plugins/ai-plugin.json (just copy this one into the plugin box)
* https://stefang.uber.space/plugins/openapi.json (this one describes how the CKAN api works)

Right now it is working poorly, but once in a while it is able to make the right calls. OA has a much weaker language model than OpenAI though, at least right now. Adapting these files for ChatGPT should be relatively easy, once they make their plugin system public.

![image](https://user-images.githubusercontent.com/5982239/236804326-a9e1cf3a-4dae-44a7-b59e-52b3e709e007.png)
(this is just a visualization, right now it hallucinates a lot of the information, even if it is capable to make the right API calls)

**Comment by stefangrotz** [comment]
Okay I did some more experiments with the plugin system. Quite often it makes the right call and then is unable to interpret it like this:

![Screenshot 2023-05-09 105438](https://user-images.githubusercontent.com/5982239/237052950-8e7adc45-77e3-48d5-8acb-de5fe2fc7248.png)

I discussed this with the developers of Open Assistant. They say to really make it work, you have to include text prompting into the API response to make it easier to understand for a language model. This means to use Plugins in CKAN a completely new API endpoint would be necessary, one that gives more context to the data. 

I don't know if the Plugin system of ChatGPT is more capable. If someone has access: it would be cool if you could test my plugin.

**Reply by GordianDziwis** [reply]
Maybe using the catalog endpoint from the [dcat](https://github.com/ckan/ckanext-dcat) plugin could help with the context?

**Reply by stefangrotz** [reply]
Thanks I will have a look and maybe experiment with it a little.

**Reply by amercader** [reply]
Can you expand or give an example on this?

> to really make it work, you have to include text prompting into the API response to make it easier to understand for a language model

what would this text prompting look like?

**Reply by stefangrotz** [reply]
This is one example API where you can find a few ideas: https://github.com/draganjovanovich/web-retriever/blob/bbac9466a8936ae84da6ad75c79a614480a6daf7/main.py#LL20C72-L20C72

It contains messages like this:

```
except Exception as e:
        print(e)
        error_message = f"Sorry, the url is not available. {e}\nYou should report this message to the user!"
        return JSONResponse(content={"error": error_message}, status_code=500)
```

But I am just learning how to do this myself.

**Comment by stefangrotz** [comment]
ChatGPT now supports custom "GPTs" with custom knowledge and capabilities: https://openai.com/blog/introducing-gpts

I used [this API definition file](https://github.com/stefangrotz/OpenDataGermanyGPT/blob/main/actions/transparenzportal-karlsruhe-ckan.json) to access the Open Data Portal Karlsruhe, and it worked quite well:
![opendata-ka](https://github.com/ckan/ckan/assets/5982239/b7cb069d-bddf-468f-9d0a-ed0729c23e2c)

**Reply by thegostev** [reply]
Thanks for keeping the ball in the air @stefangrotz it looks interesting.

**Reply by stefangrotz** [reply]
If anyone has acces to ChatGPT pro you can test it here: https://chat.openai.com/g/g-uiJQsroGg-open-data-portals-germany
I think I will create an experimental GPT for German Open Data portals. The main issue right now is that it can't handel big responses when the list is too long.
I am going to document my work here: https://github.com/stefangrotz/OpenDataGermanyGPT Feel free to join the discussion!
